iohk-monitoring/src/Control/Tracer/Transformers/Synopsizer.lhs

```diff
@@ -0,0 +1,95 @@
+\label{code:Control.Tracer.Transformers.WithThreadAndTime}
+
+%if style == newcode
+\begin{code}
+{-|
+Module: Synopsizer
+
+Synopsize runs of repeated events, as an initial message with a summary at the end.
+-}
+{-# LANGUAGE BangPatterns   #-}
+{-# LANGUAGE LambdaCase     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE MultiWayIf     #-}
+{-# LANGUAGE RankNTypes     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Control.Tracer.Transformers.Synopsizer
+    (
+    -- * transformer
+      mkSynopsizer
+    ) where
+
+import           Control.Monad.IO.Class (MonadIO (..))
+import           Data.IORef (newIORef,readIORef, writeIORef)
+
+import           Cardano.BM.Trace
+import           Cardano.BM.Data.LogItem
+import           Cardano.BM.Data.Severity
+import           Control.Tracer (Tracer (..), traceWith)
+
+\end{code}
+%endif
+
+\begin{code}
+-- | State required to track and summarise repeats.
+data SynopsizerState a
+  = SynopsizerState
+    { ssRepeats  :: {-# UNPACK #-} !Int
+    , ssLast     :: !(Maybe (LogObject a))
+    }
+
+-- | Generic Trace transformer.  Will omit consequent runs of similar log objects,
+--   with similarity defined by a supplied predicate.  When a run of similar log
+--   objects is terminated by a dissimilar one, a summary of omitted objects is
+--   emitted as a log object with same severity and privacy as the first object.
+--
+--   Handy predicates to use:  'loTypeEq' and 'loContentEq'.
+--
+--   Caveat 1:  the supplied criterion is expected to define a transitive relation.
+--
+--   Caveat 2:  the resulting trace has state, which is lost upon trace's termination.
+--   This means that if the trace ends with a run of similar messages, this will
+--   not be reflected in the trace itself, as observed by the backends
+--   -- they'll only receive the first message of the run.
+--
+--   WARNING:  the resulting tracer is not thread-safe!
+mkSynopsizer :: forall m a
+              . (MonadIO m)
+             => (LogObject a -> LogObject a -> Bool)
+             -> Trace m a -> m (Trace m a)
+mkSynopsizer matchTest tr =
```

Comment:
This is too concrete.

I suggest a more generic approach where the type signature here would be something like
`(a -> a -> Bool)  -> Trace m (Synopsis a) -> m (Trace m a)`
given
`data Synopsis a = Many !Int a | One a`
The `LogObject` version would then be a specialisation of this.

Reply:
Also, 
if there is (very long) sequence of such messages that don't change do you think that there should be a parameter that emits something every n?

Reply:
@njd42, this is specific to `Trace`'s (i.e. `Tracer`s of `LogObjects`), because we need to inject messages into the stream -- and I don't think there is a way to do this for fully polymorphic `Tracer`s?

Reply:
Regarding the periodic intermediate summarisation -- fair point, and it was mentioned in the issue as well.

Reply:
Regarding further generalisation -- I need to stretch my brain a bit on that one..

Reply:
Ok, implemented intermediate summarisation -- and I've got an idea of what the generalisation should look like.

Reply:
@deepfire Can you add an additional property test like the one you've already added for the more generalized test case once you've made those changes?

Reply:
Just pushed an implementation of generalisation.

Reply:
@oneEdoubleD, the property being tested is the same -- the transformer layer doesn't change it, really.